update multi-tenancy content (#156)
ConsenSys/doc.goquorum: docs/Concepts/Multitenancy.md

# Multi-tenancy via multiple private states

In a typical GoQuorum network, each participant (tenant) uses its own GoQuorum and Tessera node.
Tessera can be configured to manage multiple key pairs owned by one tenant.
This model is costly to run and scale as more tenants join the network.

[Use multi-tenancy via multiple private states (MPS)](../HowTo/Use/Multitenancy/Multitenancy.md) to allow multiple
tenants to use the same GoQuorum node, with each tenant having its own private state(s).
Each tenant can perform all operations (create, read, and write) on any contract in its private state, and a single
tenant can have access to [multiple private states](#multiple-private-states).

> Multi-tenancy enables a user experience similar to a user running their own managed node.

Reviewer: ```suggestion
Multi-tenancy provides a user experience similar to a user running their own managed node.
```

The public state remains publicly available to all tenants, and private states are logically separated.

## Architecture

![Multi-tenancy](../images/Multi-tenancy.png)

In this example architecture, an organization represents a tenant with multiple departments, and users within the departments.
Each tenant operates on its own private state.

> Each user in an organization owns one or more privacy manager key pairs that allows them to operate on their

Reviewer 2: ```suggestion
Each user in an organization owns one or more privacy manager key pairs that allow them to operate on their
```

(using an `Authorization` HTTP request header) when calling an API.
Calls to the GoQuorum API without an authenticated token are rejected.

The value of the scope encoded in an access token (in case of JWT), or Introspection response (in the case of the OAuth2
Token Introspection API) contains the [RPC scope](../Reference/Plugins/Security.md#oauth2-scopes) and tenant scope,
which has the following URL-based syntax:

```text
"psi://[PSI]?self.eoa=0x0&node.eoa=0x0"
```

In the syntax, `PSI` is the URL-encoded value of the PSI.

For example, for a client that has access to two private states `PS1` and `PS2`, an authorization server operator sets
up and grants the following scopes to the client:

```text
"psi://PS1?self.eoa=0x0&node.eoa"
"psi://PS2?self.eoa=0x0&node.eoa"
```

A client presenting an access token containing these scopes has full access (read/write/create) to private contracts on
private states `PS1` and `PS2` using any Ethereum accounts.

## Multiple private states

Multiple private states (MPS) is a feature that allows a GoQuorum node to manage more than one private state.
This functionality lays the foundation for multi-tenancy.

> You can [upgrade your existing GoQuorum and Tessera nodes to enable MPS](../HowTo/Use/Multitenancy/Migration.md).

Reviewer: Maybe we should let users know what versions they require? So maybe add something like, "If running GoQuorum <version> or earlier or Tessera <version> and earlier, upgrade to a later version to use the MPS functionality."

Or should we assume they are using the "latest" or "stable" version of the products, in which case we don't mention that they need to upgrade at all?

At each block:

1. All affected private states are updated and their new root hashes are calculated.
1. The trie is updated with the new private state root hashes at their PSIs.
1. A new root of the trie of private states is calculated and mapped to the public block hash.

### Private state manager

The GoQuorum private state manager resolves the [PSI](#private-state-identifier) based on input parameters.

Any RPC API call must be accompanied by a PSI or authorization token.
From the token, the private state manager derives the private state the user is attempting to access.

### Tessera resident groups

MPS uses [Tessera resident groups](https://docs.tessera.consensys.net/en/stable/Concepts/Privacy-Groups/) to map tenants
to private states.
During Tessera startup, `residentGroups` are validated to check each Tessera key is part of a single resident group.
Every Tessera key must be in a resident group for Tessera to start.
During GoQuorum startup, the `residentGroups` are [retrieved from Tessera](#tessera-q2t-communication-changes) and kept
in memory in GoQuorum.

### Tessera Q2T communication changes

> MPS introduces a new QT2 endpoint `/groups/resident` to return the resident groups defined in Tessera.

Reviewer: ```suggestion
The `/groups/resident` QT2 endpoint returns the resident groups defined in Tessera.
```

This endpoint is invoked at GoQuorum startup to retrieve all resident groups.
These details are kept in memory in GoQuorum, so the [private state manager](#private-state-manager) can resolve these
resident groups to the corresponding private state.

### Accessing a private state

Users must specify the private state they wish to operate on.

> For backwards compatibility, if a user connects without specifying the private state, the default "private" identifier is used.

Reviewer: Should "private" be `private`?

> If a "private" state is not configured, the user operates on an empty read-only private state.

Reviewer: ```suggestion
If a private state is not configured, the user operates on an empty read-only private state.
```

> In order to specify a private state to operate on the user has three options, in order of precedence:

Reviewer: ```suggestion
In order of precedence, users have the following three options to specify a private state to operate on:
```
